site/_posts/2021-03-08-establishing-reliability-as-code-via-sonobuoy.md

```diff
@@ -0,0 +1,112 @@
+
+### Getting Started
+With the Reliability Scanner, you can add new reliability checks easily.  Please refer to the [README](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/README.md) for more details.
+
+First, make sure that [Sonobuoy CLI](https://github.com/vmware-tanzu/sonobuoy), Docker, make, ytt, and kubectl are installed on your local machine.  There’s a helpful [Makefile](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/Makefile) that handles the necessary setup and templating needed to run the customized reliability check YAML.  
+
+
+### Adding a new QOS reliability check
+The reliability scanner is ready to be used to scan your cluster. However, a cluster operator who is interested in creating a customized reliability check, will need to make modifications in two places in order to create a new customized reliability check.  First, the new check must be defined by name, description, kind, and spec, in a [YAML configuration file](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/plugin/reliability-scanner-custom-values.lib.yml) as shown below:
+
+```
+name: ”Pod QOS check”
+   description: Checks each pod to see if the minimum desired QOS is defined
+   kind: v1alpha1/pod/qos # location of the check logic
+   spec: # configurable by cluster operators
+     minimum_desired_qos_class: Guaranteed # defines minimum expected QoS
+     include_detail: true # includes the actual Pod QoS in the report
+```
+
+Next, some Golang code must be written which actually does the check we need.  As a starting point, the check will have to satisfy the [Querier](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/api/v1alpha1/pod/qos/qos.go) Go interface ( the project comes with a default [QoS implementation](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/api/v1alpha1/pod/qos/qos.go) of the Querier interface), and a mapping must be made in the [scanner.go](https://github.com/vmware-tanzu/sonobuoy-plugins/blob/master/reliability-scanner/cmd/reliability-scanner/scanner.go) file.
+
+### Testing the new check
+The QoS check, defined above, allows the scanner to look across the cluster to report back on the current state of a Pod within the cluster to understand workloads that do not define our minimum [QoS class](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/).
+
+Let us create a Pod with a guaranteed QoS class and run our scan, we should see it reflected in the report:
+
+```
+$ cat <<EOF | kubectl apply -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test
+spec:
+  containers:
+  - name: test
+    image: nginx
+    resources:
+      limits:
+        memory: "200Mi"
+        cpu: "100Mi"
+EOF
+pod/test created
+```
```

Comment:
separate the command from the response using different codeblocks to avoid confusion, something like:

> ```
> ...
> EOF
> ```
> 
> we can expect the following response from `kubectl`:
> 
> ```
> pod/test created
> ```